## Script to read in clusters tsv file generated by cmonkey-python and convert it to a
## cMonkey environment with the same structure as in the original cMonkey R package.
## Requires cMonkey R package installed.
## Requires the tsv file and a tsv file for the expression data from which the clusters were generated.
## TBD: right now it re-initializes the environment each time, using the cMonkey R functions that
##   read in genome information, etc. This can be streamlined?
## Requires progs/mast in current directory (to scan the gene promoters for input motifs).

require(cMonkey)
debug.on()

require(parallel)
options(mc.cores=8)

## defaults organism, clusts.file, ratios (can be overridden before calling this script)
if ( ! exists('organism') ) organism = 'eco'
if ( ! exists('clusts.file') ) clusts.file = bzfile('eco-out-001/cmresults-postproc.tsv.bz2')
if ( ! exists('ratios') ) ratios <- 'DISTILLER_data.tsv'
if ( ! exists('n.motifs') ) n.motifs <- 2
if ( ! exists('force.motif') ) force.motif <- FALSE ## force it to re-motif rather than using meme results from tsv

x=read.delim(clusts.file)
#for (i in c('resid','dens_string','meanp_meme')) x[[i]] = as.numeric(gsub('f0','',as.character(x[[i]])))
e=cmonkey.init(organism=organism, k.clust=nrow(x), seed.method=c(rows='rnd',cols='rnd'), ratios=ratios,
  discard.genome=F, parallel.cores=options('mc.cores')[[1]], parallel.cores.motif=options('mc.cores')[[1]])
e$cmonkey.re.seed( e )
#sys.source("~/scratch/biclust/cmonkey-funcs.R",envir=e,chdir=T)
e$row.score.func='default'
e$iter = 1999
seq.type = 'upstream meme'
e$mast.cmd[1]='./progs//mast $memeOutFname -d $fname -bfile $bgFname -nostatus -stdout -text -brief -ev 99999 -mev 99999 -mt 0.99 -seqp -remcorr'
names(e$ratios)='ratios' ## for some reason gets set to 'ratios.1'
all.seqs = e$genome.info$all.upstream.seqs[[ seq.type ]]
bg.list = e$genome.info$bg.list[[ seq.type ]]
bg.fname = e$genome.info$bg.fname[ seq.type ]

##for (i in 1:nrow(x)) {
tmp = lapply(1:nrow(x), function(i){
  cat(i," ",x$k[i],"\n")
  rows=strsplit(as.character(x$rows[i]),',')[[1]]
  cols=strsplit(as.character(x$cols[i]),',')[[1]]
  clust = e$clusterStack[[i]]
  clust$rows = rows
  clust$cols = cols
  clust$resid=x$resid[i]
  pv.ev = list()
  e$clusterStack[[i]] = clust

  motif.out <- list()
  if ( ! force.motif ) {
    meme.out.ORIG = strsplit(as.character(x$meme_out[i]),"<<<<>>>>")[[1]]
    if ( length( meme.out.ORIG ) > 0 ) meme.out = e$getMemeMotifInfo( meme.out.ORIG )
    else meme.out = NULL
  
    mast.out <- NULL
    pv.ev = list()
    if ( ! is.null( meme.out ) ) {
      mast.out <- try( e$runMast( meme.out.ORIG, e$mast.cmd[ seq.type ], names( all.seqs ), all.seqs,
                                 verbose=TRUE, seq.type=seq.type, bg.list=bg.list, bgfname=bg.fname ) ) 
      if ( class(mast.out) != 'try-error' ) pv.ev <- e$get.pv.ev.single( mast.out, rows )
    }
    motif.out$meme.out <- meme.out
    motif.out$pv.ev <- pv.ev
    motif.out$k <- i; motif.out$last.run <- FALSE; motif.out$prev.run <- NULL

  } else {
    meme.out <- e$motif.one.cluster( i, seq.type=names(e$meme.cmd)[1], verbose=T, force=T )
    motif.out <- meme.out
  }

  ##e$meme.scores[[seq.type]][[i]] = meme.out
    
  if ( length(meme.out) > 0 ) {
    ##e$clusterStack[[i]] = clust
  }

  list( meme.out=motif.out, clust=clust )
} )

for (i in 1:nrow(x)) {
  e$meme.scores[[seq.type]][[i]] = tmp[[i]]$meme.out
  clust <- tmp[[i]]$clust
  e$clusterStack[[i]] <- clust
  tmp2 = e$cluster.pclust(i)
  clust$e.val <- tmp2$e.vals
  clust$p.clust <- tmp2$p.clusts
  e$clusterStack[[i]] = clust
}

rm(tmp, tmp2); gc()

e$meme.scores[[seq.type]]$all.pv <- e$make.pv.ev.matrix( e$meme.scores[[seq.type]] )

save( e, file=sprintf('%s_out.RData', organism) )
if ( exists( 'out.dir' ) ) system( sprintf( 'mv -v %s_out.RData %s', organism, out.dir ) )

if ( FALSE && organism == 'eco' ) {
  cwd <- setwd('cmEvalEco')
  try(capture.output(try(source("compare_regulondb_single.R",local=T))))

  use.combi <- TRUE
  source('compare_regdb_regulons.R',local=T)
  source("print_stats.R",local=T)

  use.combi <- FALSE
  source('compare_regdb_regulons.R',local=T)
  source("print_stats.R",local=T)
  
  setwd(cwd)

  save.image( file=sprintf('%s_out_with_regdb_comparison.RData', organism) )
  if ( exists( 'out.dir' ) ) system( sprintf( 'mv -v %s_out_with_regdb_comparison.RData %s', organism, out.dir ) )
}

